README.md
# CSA
Comprehensive System Accounting for Linux

### What is Comprehensive System Accounting?
CSA is a more extensive alternative to existing Linux process accounting packages. It exists as a set of C programs and shell scripts that provide methods for collecting per-process resource usage data, monitoring disk usage, and charging fees to specific login accounts. CSA takes this per-process accounting information and combines it by job identifier (jid) for tracking groups of related processes within system boot uptime periods.

CSA provides the following features which are not available with other Linux accounting packages:

- Per-job accounting
- Daemon accounting
- User job accounting (`ja` command)
- Flexible accounting periods (not just daily and monthly periods)
- Flexible system billing units (SBUs)
- Offline archiving of accounting data
- User exits for site specific customization of reports
- Configurable parameters

How else is this different from other Linux accounting packages? Linux has a form of process accounting which is similar to SysV/BSD basic process accounting. The set of Linux accounting commands provide similar functionality but have different names and options.

While the Linux kernel itself monitors the size of the accounting records file (pacct), CSA monitoring of file size takes place outside of the kernel.

Linux CSA is the outcome of a joint effort between Los Alamos National Laboratory (LANL) and SGI to port the IRIX CSA tool to the Linux Operating System. This project is a community driven fork of that code base and is not supported by either organization.

### More Info
For additional information and instructions on proper usage, please visit the  [Wiki](https://github.com/LinuxCSA/CSA/wiki).


csa/ChangeLog
* Fri Aug 09 2019 Jay Lan <jay.j.lan@nasa.gov>
- Upgraded CSA codes to new netlink API libnl3-200-3.3.0.

* Fri Aug 09 2019 Jay Lan <jay.j.lan@nasa.gov>
- Fixed Warning messages produced by gcc during compilation.
- gcc-7-1.563.x86_64

* Tue May 30 2017 Jay Lan <jay.j.lan@nasa.gov>
- Fixed an issue that csad and jobd sometimes get out of sync.

* Wed Jun  8 2016 Jay Lan <jay.j.lan@nasa.gov>
- Converted csa package from the last SGI-supported version 4.1.1 to
  community version 5.0.0

* Fri Oct 17 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.1
- Various fixes to allow for SLES11 build

* Fri Jul 25 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.1.0
- Fixed memory leaks
- Changed acctjob structure offset type from int to int64_t

* Fri Apr 18 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.2
- Updated to use new libnl API
- Fixed process exit code assignment

* Tue Mar 4 2008 Jonathan Lim <jlim@sgi.com>
- csa-4.0.1
- Cleaned up acctdusg
- Added check to prevent SIGPIPE when communicating with daemon
- Increased maximum PID allowed for larger systems

* Wed Nov 7 2007 Jonathan Lim <jlim@sgi.com>
- csa-4.0.0
- Implement kernel functionality as a userspace daemon
- Remove obsolete wait metrics and replace with taskstats delay accounting
- Various bug fixes

* Fri Apr 27 2007 Jay Lan <jlan@sgi.com>
* Mon Apr 30 2007 Jonathan Lim <jlim@sgi.com>
- csa-3.0.2
- SBU's data was all zero in csa-3.0.0 and csa-3.0.1
- Clarify csacom(1) man page on '-f' option
- Obsolete csacom '-a' & '-q' option
- Fix a problem of csadrep reading input from a binary file
- Fixed csadrep(8) man page: replaced '-w' with '-n' and removed '-t'

* Wed Jul 5 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.1
- 'csacom -b' was broken in csa-3.0.0
- 'csacom -b' drops the last record of a pacct file

* Thu Feb 24 2006 Jay Lan <jlan@sgi.com>
- csa-3.0.0
- csacom -[esES] option was broken.
- create /usr/sbin/rccsa symlink at %install in the spec file
- remove code for CSA record types not supported in Linux
- spliting up csa_kern.h into kernel version and userland version
- support pacct files created in csa-2.*.*

* Wed Jul 27 2005 Chris Sturtivant <csturtiv@sgi.com>
- csa-2.2.2
- Disabled perror for permission denied in csa_ctl when called by csa_auth

* Mon May 2 2005 Jay Lan <jlan@sgi.com>
- csa-2.2.1
- clean up csa init script
- csa_wracct(3) man page update
- 'csacom -o' wrote only one record
- 'csacom -b' does not work
- csajrep man page references invalid option '-W'
- 'csajrep -r' does not work
- 'dodisk' created files with incorrect user and group name 'adm'
- CSA does not handle gid's > 65535
- 'csacom -j' failed if job id greater than max value of an int

* Mon Nov 22 2004 Jay Lan <jlan@sgi.com>
- csa-2.2.0
- Fixed compilation issue with gcc 3.3.3
- CSA assumes existence of 'adm' user/group account
- csacom & csacms usage otuput missing carriage return
- csarecy misses a carriage return on certain condition
- csacom -W segfaults after printing one bad time

* Fri Oct 8 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.2
- Fixed 'csaswitch -c halt' bug
- Fixed 'csaedit dumped core on bad wkmg record' bug

* Tue Jul 27 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.1
- Remove static API library libcsa.a from rpm package
- Fixed a bug at csacom with invalid -u option
- cascom headers and data not aligned correctly at output

* Wed Jun 30 2004 Jay Lan <jlan@sgi.com>
- csa-2.1.0
- Fixed a problem dealing with accounting files > 2GB
- Now derives "Day of Year" from other fields of the holidays file
- Removed "Pages Swapped" from csacom and csacms - not supported in Linux
- Fixed a bug of not mapping UID 0 to root

* Tue Feb 17 2004 Jay Lan <jlan@sgi.com>
- csa-2.0.0
- Implemented /proc ioctl interface
- Renamed rpm from csa-job-acct to csa
- Renamed /etc/init.d/csaacct to /etc/init.d/csa

* Fri Nov 7 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-2.0.0
- Provided API library libcsa.so and associated header file

* Wed Sep 3 2003 Jay Lan <jlan@sgi.com>
- csa-job-acct-1.0.3
- Updated the spec file

* Thu Feb 13 2003 Robin Holt <holt@sgi.com>
- Updated init scripts to use /var/lock/subsys

* Tue Nov 5 2002 Richard Offer <offer@sgi.com>
- Fixed packager/distribution/vendor

* Thu May 23 2002 Robin Holt <holt@sgi.com>
- csa-job-acct-1.0.2
- Changed install destination from /usr/local to /usr
- Updated the man page location to /usr/share/man
- Added a clean and realclean target to the Makefile

* Thu Apr 5 2001 Marlys Kohnke <kohnke@sgi.com>
- Changed csadrep internal markers back to match those in IRIX CSA
- Changed ja summary report so that all lines are displayed even if the
  resource used values are zero

* Sun Feb 11 2001 Marlys Kohnke <kohnke@sgi.com>
- Added support for Workload Management daemon accounting records
- Fixed csachkpacct script problem (df -k needs to be df -kP; changed awk
  field to 4)

* Fri Jan 19 2001 Daryl Grunau <dwg@acl.lanl.gov>
- port to Linux from SGI IRIX
- csa-job-acct-1.0.0 (available January 19, 2001)
- initial csa-job-acct-1.0.0 open source release


job/cmd/jkill.c
/*
 * Copyright (c) 2000-2007 Silicon Graphics, Inc. 
 * All Rights Reserved.
 *
 */

/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation; either version 2.1 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 */


#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <job.h>
#include <ctype.h>
#include <strings.h>

struct sigmap {
	char *name;
	int  number;
};

struct sigmap sigmap[] = {
		/* POSIX signals */
	{ "HUP",	SIGHUP },	/* 1 */
	{ "INT",	SIGINT }, 	/* 2 */
	{ "QUIT",	SIGQUIT }, 	/* 3 */
	{ "ILL",	SIGILL }, 	/* 4 */
	{ "ABRT",	SIGABRT }, 	/* 6 */
	{ "FPE",	SIGFPE }, 	/* 8 */
	{ "KILL",	SIGKILL }, 	/* 9 */
	{ "SEGV",	SIGSEGV }, 	/* 11 */
	{ "PIPE",	SIGPIPE }, 	/* 13 */
	{ "ALRM",	SIGALRM }, 	/* 14 */
	{ "TERM",	SIGTERM }, 	/* 15 */
	{ "USR1",	SIGUSR1 }, 	/* 10 (arm,i386,m68k,ppc), 30 (alpha,sparc*), 16 (mips) */
	{ "USR2",	SIGUSR2 }, 	/* 12 (arm,i386,m68k,ppc), 31 (alpha,sparc*), 17 (mips) */
	{ "CHLD",	SIGCHLD }, 	/* 17 (arm,i386,m68k,ppc), 20 (alpha,sparc*), 18 (mips) */
	{ "CONT",	SIGCONT }, 	/* 18 (arm,i386,m68k,ppc), 19 (alpha,sparc*), 25 (mips) */
	{ "STOP",	SIGSTOP },	/* 19 (arm,i386,m68k,ppc), 17 (alpha,sparc*), 23 (mips) */
	{ "TSTP",	SIGTSTP },	/* 20 (arm,i386,m68k,ppc), 18 (alpha,sparc*), 24 (mips) */
	{ "TTIN",	SIGTTIN },	/* 21 (arm,i386,m68k,ppc,alpha,sparc*), 26 (mips) */
	{ "TTOU",	SIGTTOU },	/* 22 (arm,i386,m68k,ppc,alpha,sparc*), 27 (mips) */
	/* Miscellaneous other signals */
#ifdef SIGTRAP
	{ "TRAP",	SIGTRAP },	/* 5 */
#endif
#ifdef SIGIOT
	{ "IOT",	SIGIOT }, 	/* 6, same as SIGABRT */
#endif
#ifdef SIGEMT
	{ "EMT",	SIGEMT }, 	/* 7 (mips,alpha,sparc*) */
#endif
#ifdef SIGBUS
	{ "BUS",	SIGBUS },	/* 7 (arm,i386,m68k,ppc), 10 (mips,alpha,sparc*) */
#endif
#ifdef SIGSYS
	{ "SYS",	SIGSYS }, 	/* 12 (mips,alpha,sparc*) */
#endif
#ifdef SIGSTKFLT
	{ "STKFLT",	SIGSTKFLT },	/* 16 (arm,i386,m68k,ppc) */
#endif
#ifdef SIGURG
	{ "URG",	SIGURG },	/* 23 (arm,i386,m68k,ppc), 16 (alpha,sparc*), 21 (mips) */
#endif
#ifdef SIGIO
	{ "IO",		SIGIO },	/* 29 (arm,i386,m68k,ppc), 23 (alpha,sparc*), 22 (mips) */
#endif
#ifdef SIGPOLL
	{ "POLL",	SIGPOLL },	/* same as SIGIO */
#endif
#ifdef SIGCLD
	{ "CLD",	SIGCLD },	/* same as SIGCHLD (mips) */
#endif
#ifdef SIGXCPU
	{ "XCPU",	SIGXCPU },	/* 24 (arm,i386,m68k,ppc,alpha,sparc*), 30 (mips) */
#endif
#ifdef SIGXFSZ
	{ "XFSZ",	SIGXFSZ },	/* 25 (arm,i386,m68k,ppc,alpha,sparc*), 31 (mips) */
#endif
#ifdef SIGVTALRM
	{ "VTALRM",	SIGVTALRM },	/* 26 (arm,i386,m68k,ppc,alpha,sparc*), 28 (mips) */
#endif
#ifdef SIGPROF
	{ "PROF",	SIGPROF },	/* 27 (arm,i386,m68k,ppc,alpha,sparc*), 29 (mips) */
#endif
#ifdef SIGPWR
	{ "PWR",	SIGPWR },	/* 30 (arm,i386,m68k,ppc), 29 (alpha,sparc*), 19 (mips) */
#endif
#ifdef SIGINFO
	{ "INFO",	SIGINFO },	/* 29 (alpha) */
#endif
#ifdef SIGLOST
	{ "LOST",	SIGLOST }, 	/* 29 (arm,i386,m68k,ppc,sparc*) */
#endif
#ifdef SIGWINCH
	{ "WINCH",	SIGWINCH },	/* 28 (arm,i386,m68k,ppc,alpha,sparc*), 20 (mips) */
#endif
#ifdef SIGUNUSED
	{ "UNUSED",	SIGUNUSED },	/* 31 (arm,i386,m68k,ppc) */
#endif
};

int sigmapsz = sizeof(sigmap)/sizeof(sigmap[0]);

void usage(void) 
{
	fprintf(stderr, "usage: jkill [-l | [ [-signal] jid ... ] ]\n");
}

int sig_name2number(char *name)
{
	int i;

	/* 
	 * If signal name string is prepended by SIG/sig, skip the
	 * first 3 characters.
	 */
	if (!strncasecmp("sig", name, 3)) 
		name += 3;

	/* 
	 * Match the string against those in the signal map table,
	 * return the corresponsing signal number or -1 if signal not
	 * found.
	 */
	for (i = 0; i < sigmapsz; i++) {
		if (!strcasecmp(sigmap[i].name, name))
			return sigmap[i].number;
	}
	return -1;
}

int list_signals(void)
{
	int i; 
	int col = 0;

	for (i = 0; i < sigmapsz; i++) {
		printf("SIG%-7s[%2d]", sigmap[i].name, sigmap[i].number);
		if (++col == 4) {
			printf("\n");
			col = 0;
		} else { 
			printf("      ");
		}
	}
	printf("\n");
	return 0;
}


int
main(int argc, char *argv[])
{
	int argcnt = argc;
	int sig_number = SIGTERM; /* default, use if no signal specified */
	int list = 0;
	char *arg;
	jid_t jid;
	int retval;


	/* Parse the command line string for options */
	for (argc--, argv++; argc > 0; argc--, argv++) {
		arg = *argv;

		if (arg[0]!= '-') {
			/* no options, only jid values */
			break;
		}

		/* Check option for signal number */
		if ((sig_number = sig_name2number(&arg[1])) < 0) {
			/* Check if option is signal number */
			if (isdigit(arg[1])) {
				sig_number = atoi(&arg[1]);
			}

			/* option not signal, check for list option */
			else if (arg[1] == 'l' && arg[2] == '\0') {
				/* list option must be only argument */
				if (argcnt != 2) {
					usage();
					exit(1);
				} else {
					list_signals();
					return 0;
				}
			} 
			
			/* option not signal or list, so error */
			else {
				usage();
				exit(1);
			}
		}
	}

	/* process list of jid value & send signals */
	for ( ; argc > 0; argc--, argv++) {
		sscanf(*argv, "%Lx", &jid);

		retval = job_killjid(jid,sig_number);
		if (retval == -1) {
			if (errno == ENOENT) {
				fprintf(stderr, "jobd does not appear to be running.\n");
			} else if (errno == ENODATA) {
				fprintf(stderr, "No such job\n");
			} else {
				perror("job_killjid");
			}
			exit(1);
		}
	
		printf("Processes in job[%0#18Lx] sent signal %d.\n", 
				jid, sig_number);
	}
	
	exit(0);
}
